src/backend/distributed/utils/query_utils.c

```diff
@@ -0,0 +1,109 @@
+/*-------------------------------------------------------------------------
+ *
+ * query_utils.c
+ *
+ * Query-walker utility functions to be used to construct a logical plan
+ * tree from the given query tree structure.
+ *
+ * Copyright (c), Citus Data, Inc.
+ *
+ ****-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "nodes/nodeFuncs.h"
+#include "distributed/query_utils.h"
+#include "distributed/query_pushdown_planning.h"
+
+/* This function will be used to walk over a tree to gather entries.
+ * Execution is parameterized by passing walkerMode flag via ExtractRangeTableWalkerContext
+ * as we cannot pass more than one parameter to query_tree_walker
+ */
+bool
+ExtractRangeTableList(Node *node, ExtractRangeTableWalkerContext *context)
+{
+	/* get parameters from context */
+	List **rangeTableRelationList = context->rangeTableList;
+	ExtractRangeTableMode walkerMode = context->walkerMode;
+
+	bool walkIsComplete = false;
+
+	if (node == NULL)
+	{
+		return false;
+	}
+
+	if (IsA(node, RangeTblEntry))
+	{
+		RangeTblEntry *rangeTable = (RangeTblEntry *) node;
+
+		if (walkerMode == EXTRACT_ALL_ENTRIES ||
+			(rangeTable->rtekind == RTE_RELATION && rangeTable->relkind != RELKIND_VIEW))
```

Comment:
I like the refactoring. But, what about `RELKIND_MATVIEW`? I think we should probably skip that as well as `RELKIND_VIEW ` ?

In general, we seem to be only interested** in `RELKIND_RELATION`, maybe we  should change the if statement to  `rangeTable->rtekind == RTE_RELATION && rangeTable->relkind == RELKIND_RELATION)`. It also makes the `EXTRACT_RELATION_ENTRIES ` flag fit into its name better.

In case anyone interested in other `relkind`s, they can add `EXTRACT_` flags on top of this?

** check-full passes with this [diff](https://github.com/citusdata/citus/compare/onder_test_2) , besides I don't see any code that'd need the other `relkind`s explicitly in Citus

Reply:
That makes sense, working on the changes